### PR TITLE
Hardfork import integration test

### PR DESCRIPTION
**Pull request by sasurobert**
Small fixes and integration test for export and import.

### CODE REVIEW

File: integrationTests/multiShard/hardFork/hardFork_test.go [hunk 1]
```diff
@@ -180,27 +178,122 @@ func TestEpochStartChangeWithContinuousTransactionsInMultiShardedEnvironment(t *
...
+		require.Nil(t, err)
+		require.NotNil(t, genesisBlocks)
+
+		node.GenesisBlocks = genesisBlocks
+		for _, genesisBlock := range genesisBlocks {
+			log.Info("hardfork genesisblock roothash", "shardID", genesisBlock.GetShardID(), "rootHash", genesisBlock.GetRootHash())
+		}
+	}
+
+	for _, nodeA := range nodes {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/multiShard/hardFork/hardFork_test.go:261]: made a spearate function

File: update/errors.go [hunk 2]
```diff
@@ -197,5 +194,5 @@ var ErrNilPendingTxProcessor = errors.New("nil pending tx processor")
 // ErrNilHardForkBlockProcessor signals that nil hard fork block processor has been provided
 var ErrNilHardForkBlockProcessor = errors.New("nil hard fork block processor")
 
-// ErrNilTrieFactory signals that nil trie factory has been provided
-var ErrNilTrieFactory = errors.New("nil trie factory")
+// ErrNilTrieStorageManager signals that nil trie storage manager has been provided
+var ErrNilTrieStorageManager = errors.New("nil storage manager has been factory")
```
Comment (SebastianMarian) [inline update/errors.go:198]: Refactor this error message: -> "nil trie storage manager"

File: update/files/multiFileReader.go [hunk 3]
```diff
@@ -100,11 +100,13 @@ func (m *multiFileReader) ReadNextItem(fileName string) (string, []byte, error)
 	}
 
 	hexEncoded := scanner.Text()
+
 	key, err := hex.DecodeString(hexEncoded)
 	if err != nil {
 		return "", nil, err
 	}
 
+	log.Trace("import key", "key", string(key))
```
Comment (SebastianMarian) [inline update/files/multiFileReader.go:109]: "import" is enough as "key" is printed after

File: update/files/multiFileWriter.go [hunk 4]
```diff
@@ -90,6 +90,7 @@ func (m *multiFileWriter) Write(fileName string, key string, value []byte) error
 		return err
 	}
 
+	log.Trace("export key", "key", key)
```
Comment (SebastianMarian) [inline update/files/multiFileWriter.go:93]: "export" is enough as "key" is printed after

File: update/genesis/import.go [hunk 5]
```diff
@@ -307,7 +341,11 @@ func (si *stateImport) importState(fileName string, trieKey string) error {
 	}
 
 	oldRootHash := value
-	log.Debug("old root hash", "value", oldRootHash)
+	log.Debug("old root hash", "value", oldRootHash, "read root hash", rootHash)
```
Comment (SebastianMarian) [inline update/genesis/import.go:344]: log.Debug("root hash", "old", oldRootHash, "read", rootHash)

File: update/genesis/import.go [hunk 6]
```diff
@@ -307,7 +341,11 @@ func (si *stateImport) importState(fileName string, trieKey string) error {
...
+
+	if bytes.Equal(value, trie.EmptyTrieHash) {
```
Comment (SebastianMarian) [inline update/genesis/import.go:346]: You can use oldRootHash instead value, to be more clear here

File: update/genesis/import.go [hunk 7]
```diff
@@ -359,6 +398,7 @@ func (si *stateImport) saveRootHash(accountsDB state.AccountsAdapter, accType Ty
 		accountsDB.SnapshotState(rootHash)
 	}
 
+	log.Info("imported state rootHash", "rootHash", rootHash, "shID", shardID)
```
Comment (SebastianMarian) [inline update/genesis/import.go:401]: "imported state" is enough, "rootHash" is printed after

File: update/genesis/base.go [hunk 8]
```diff
@@ -102,13 +99,13 @@ func GetTrieTypeAndShId(key string) (Type, uint32, error) {
 		return UserAccount, 0, update.ErrUnknownType
 	}
 
-	accTypeInt64, err := strconv.ParseInt(splitString[2], 10, 0)
+	accTypeInt64, err := strconv.ParseInt(splitString[3], 10, 0)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply update/genesis/base.go]: Done

File: update/genesis/base.go [hunk 9]
```diff
@@ -102,13 +99,13 @@ func GetTrieTypeAndShId(key string) (Type, uint32, error) {
...
 	if err != nil {
 		return UserAccount, 0, err
 	}
-	accType := getAccountType(int(accTypeInt64))
+	accType := Type(accTypeInt64)
 
-	shId, err := strconv.ParseInt(splitString[1], 10, 0)
+	shId, err := strconv.ParseInt(splitString[2], 10, 0)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply update/genesis/base.go]: Dine

File: update/errors.go (hunk 2 above)
Comment (sasurobert) [reply update/errors.go:198]: done

File: update/files/multiFileReader.go (hunk 3 above)
Comment (sasurobert) [reply update/files/multiFileReader.go:109]: done

File: update/files/multiFileWriter.go (hunk 4 above)
Comment (sasurobert) [reply update/files/multiFileWriter.go:93]: done

File: update/genesis/import.go (hunk 5 above)
Comment (sasurobert) [reply update/genesis/import.go:344]: not needed. deleted.

File: update/genesis/import.go (hunk 6 above)
Comment (sasurobert) [reply update/genesis/import.go:346]: done

File: update/genesis/import.go (hunk 7 above)
Comment (sasurobert) [reply update/genesis/import.go:401]: done

**Comment by elrondscammer3** [comment]
That's enough!! Cheating bastards.
The price keeps falling. People don't trust you. ERD doesn't buy. What do you boys do for a living?

**Comment by elrondscammer2** [comment]
ERD/Elrond Scam Coin. Team ERD is a con man. Don't buy ERD. I send this
 message every day to everyone in the group. Prevent more people from being conned by sharing with you.
  ERD BIG DUMP SELL NOW ...

**Comment by elrondscammer2** [comment]
Beniamin Mincu 
ERD/Elrond Scam Coin. Team ERD is a con man. Don't buy ERD. I send this
 message every day to everyone in the group. Prevent more people from being conned by sharing with you.
  ERD BIG DUMP SELL NOW ...

**Comment by elrondscammer4** [comment]
ALERT! ALERT! ELROND WAS HACKED!

**Comment by elrondscammer4** [comment]
People have been waiting months for these sons of bitches to do something.

**Comment by elrondscammer4** [comment]
It's been months and the price hasn't gone up. We bought ERD for these sons of bitches?

**Pull request merged**